### DISCUSSION TITLE
Foldable editor sections for smaller screens

### DISCUSSION THREAD

**Discussion by tkgalk** [godotengine/godot-proposals#6958] [Editor]
![image](https://github.com/godotengine/godot-proposals/assets/3187536/8c711a60-ccb3-4a9b-82e3-2fa86c6c85ed)

Idea for a potential improvement for Godot Editor, especially useful for smaller screens like laptops. The ability to fold the editor windows to the side like in JetBrains editor or Unreal Engine 5. It could be incredibly powerful if keybindings could be assigned to those folded windows to toggle them back and forth.

**Comment by theraot** [comment]
Something like this has come up before... Do you think they should always fold, or do you think there should be a way to pin them? And if so, how do you propose that part of the UI to be?

---

Meanwhile, if you just want more space for the code, try Distraction Free Mode. It is the four arrows icon near the top, or you can use they keyboard shortcut which is Ctrl+Shift+F11 by default (you can edit it on the Shortcut tab of the Editor Settings window).

*If you care about the minimum width of the Godot window, like I do, I don't have a solution.*

You might also find it helpful to change the dock positions and save them (using the Editor -> Editor Layout sub-menu).

**Reply by Zireael07** [reply]
I was the one to propose folding/unfolding in one of the comments to "improve editor UX" proposals. My use case is on desktop, having a Godot window side-by-side with some other window, e.g. a tutorial or some reference code. So reducing the minimum width is a must have.

I guess they could fold if Godot window width is below some threshold. But pinning could work too (with a pin icon on that fold)

I imagined portions of the editor unfolding with either a click on the fold, or (for mobile, where the same editor width problem often applies) swiping to window edge to fold, and back towards the center to unfold.

**Reply by tkgalk** [reply]
I think when put into this "side mode", an automatic folding with an option to pin/lock would be the best.

**Reply by theraot** [reply]
If I recall correctly the devil is in the details.

Where do you put the pin icon?

- Is it next of the name of tabs?

   Would each tab have one pin? Can some docks on a position be folded while others don't?

- Is it left to the ⋮?

   How does it interact with the ⧀ and ⧁ icons that appear when there are too many docks?

- Is it inside the Is it next to the ⋮ menu?

   Perhaps "Make floating", "Pinned" and "Folding" are options. Perhaps it is too many clicks?
   
- Somewhere else?

I think once this is decided, somebody can go and do it, at least that part.

---

The minimum width of the window has an extra wrinkle:

- Should the minimum width change when all the docks are folded (or in Distraction Free Mode for that matter). *I imagine this implementation is error prone.*
- Or should reducing the width of the window automatically fold (or go into Distraction Free Mode). *Sadly, sometimes what I want to see is the inspector.*
- Should it just be reduced, and that's it? *I doubt this would be accepted since it hasn't been reduced yet.*

**Reply by tkgalk** [reply]
Good questions.

> Is it next of the name of tabs?

I think so, yes.

> Would each tab have one pin? Can some docks on a position be folded while others don't?

I think a whole "group" should have a single pin.

> How does it interact with the ⧀ and ⧁ icons that appear when there are too many docks?

They would move the tab within the same group as they do now and influence the order of the tab within a folded group as well as the order if the group is visible (precisely as they are now). Each folded group should have one tab marked as active and the rest would be inactive. When the group is opened this is the order they appear in and the active one is the one shown, with other ones behaving like tabs today.

**Reply by Zireael07** [reply]
Re: the pin - my instinct is to put it next to the name

Minimum width - your second bullet point is what I imagine automatic folding working like. If you want to see the inspector, you manually unfold and then fold manually back.

**Comment by azur-wolve** [comment]
`Ctrl + \` → toggle scripts panel
`Ctrl + Shift + F11` → toggle workspace full screen

**Comment by theraot** [comment]
I made mock ups for the pin icon position (including what I'd consider bad options).

1. No Pin Icon

    ![1](https://github.com/godotengine/godot-proposals/assets/1212509/04110a08-9b83-4f73-8de6-19a8b8f88a85)
    ![0](https://github.com/godotengine/godot-proposals/assets/1212509/635f9a09-9010-4d8e-b85e-f44ec29a09cf)

2. To the right of ⋮:

    ![1a](https://github.com/godotengine/godot-proposals/assets/1212509/f6e8c373-3cdd-4b99-9807-87d45bcef1fa)
    ![0a](https://github.com/godotengine/godot-proposals/assets/1212509/19d462a0-d41c-479c-bb79-dd459c00d995)

3. To the left of ⋮:

    ![1b](https://github.com/godotengine/godot-proposals/assets/1212509/81cc50d0-2bd4-4359-b323-dba9d7037f4a)
    ![0b](https://github.com/godotengine/godot-proposals/assets/1212509/708c9556-401e-40fe-b209-452cc5cad3db)

4. To the left of ⧀ and ⧁:

    ![1c](https://github.com/godotengine/godot-proposals/assets/1212509/c201ed67-6637-49ad-9ef3-7a50cb4c5d2a)
    ![0c](https://github.com/godotengine/godot-proposals/assets/1212509/e7a764d4-cd50-4129-87a0-7b2e99f1596a)

5. To the right of the tabs:

    ![0d](https://github.com/godotengine/godot-proposals/assets/1212509/dee3a348-74ce-4618-84ee-1ac7f58db9ce)
    ![1d](https://github.com/godotengine/godot-proposals/assets/1212509/c27b8e29-ff48-4b12-a8ed-c604aa6080a3)

6. On every tab:

    ![1e](https://github.com/godotengine/godot-proposals/assets/1212509/43d674bf-7039-4328-9d8c-87256b53757e)
    ![0e](https://github.com/godotengine/godot-proposals/assets/1212509/966cd2bc-4b60-4b20-bf95-460f84b2013e)

7. Only on the active tab:

    ![1f](https://github.com/godotengine/godot-proposals/assets/1212509/1c011ae3-5f87-4cdd-accd-f49bf3d22151)
    ![0f](https://github.com/godotengine/godot-proposals/assets/1212509/7e3cb248-c23f-4218-a090-d59e97f7eddd)

8. **Addendum**: On the left of the tabs. *Not depicted.*

Disregard margins and colors. Those could be tweaked independently.

<details><summary>My preference</summary>Option 3.</details>

@Zireael07 your take?

**Reply by Zireael07** [reply]
6 or 7, no preference for either ;)

**Reply by tkgalk** [reply]
Definitely number 3.
